Adding 'Transactions' and 'Client-side latency' sections to the troubleshooting docs (#22793)
MaterializeInc/materialize: doc/user/content/manage/troubleshooting.md

From the results of this query we can see that most of the elapsed time of the dataflow is caused by the time it takes to maintain the updates in one of the arrangements of the dataflow of the materialized view.

## Why is Materialize unresponsive or slow?
Slow responses can be the result of a number of factors, which we can help you understand and address.

If you haven't already, you should consider an [index](https://materialize.com/docs/sql/create-index/) for your query, and check out the [optimizations](https://materialize.com/docs/transform-data/optimization/) on how to optimize query performance in Materialize.

> You can also consider changing your isolation level, depending on the [consistency guarantees](https://materialize.com/docs/get-started/isolation-level) that you need.

Reviewer: This section is mixing recommendations on what to do if Materialize is unresponsive or slow, and root causes. I can push a suggestion to reorganize the content as a follow-up (wouldn't want to block your PR on that).

Author: That would be awesome! Hoping to use this as a springboard for other improvements.

### Transactions

Transactions are a database concept for bundling multiple query steps into a single, all-or-nothing operation. You can read more about them in the [transactions](https://materialize.com/docs/sql/begin) section of our docs.

In Materialize, `BEGIN` starts a transaction block. All statements in a transaction block will be executed in a single transaction until an explicit `COMMIT` or `ROLLBACK` is given. All statements in that transaction happen at the same timestamp.

Because Materialize does not know which objects (sources, indexes, or materialized views) will be queried during the transaction, the objects in the first `SELECT` and any other object in the same schemas are assumed to be possible query targets. As a result, during that first query, a timestamp for the transaction is chosen that is valid for all of the objects in that same schema.

> What this means for latency: even in serializable isolation level, some queries made within a transaction may not be able return immediately. For example, if a materialized view is lagging or a source is stalled out, a query on that object in a transaction will block until the object catches up to the current time.

Reviewer 2: Can you link "serializable isolation level" to the serializable section of the isolation level docs?

> What this means for latency: even in serializable isolation level, some queries made within a transaction may not be able return immediately. For example, if a materialized view is lagging or a source is stalled out, a query on that object in a transaction will block until the object catches up to the current time.

Reviewer 3: I think you could replace this block with (stealing from your text):

>  In Materialize, `BEGIN` starts a transaction block. All statements in a transaction block will be executed in a single transaction until an explicit `COMMIT` or `ROLLBACK` is given. All statements in that transaction happen at the same timestamp, and that timestamp must be valid for all objects the transaction may access. Materialize may delay queries against "slow" tables until they catch up to faster tables in the same schema. We recommend you avoid using transactions in contexts where you require low latency responses and are not certain that all tables in a schema will be equally current.

Author: I like this a lot, thanks Frank!

> What this means for latency: even in serializable isolation level, some queries made within a transaction may not be able return immediately. For example, if a materialized view is lagging or a source is stalled out, a query on that object in a transaction will block until the object catches up to the current time.

Reviewer: ```suggestion
What this means for latency: even in [serializable isolation level](/get-started/isolation-level/#serializable), some queries made within a transaction may not be able return immediately. For example, if a materialized view is lagging or a source is stalled out, a query on that object in a transaction will block until the object catches up to the current time.
```

What you can do:

> - Avoid using transactions where you don’t need them, if you’re only executing single statements at a time, for example.

Reviewer 2: ```suggestion
- Avoid using transactions where you don’t need them. For example, if you’re only executing single statements at a time.
```

- Double check whether your SQL library or ORM is wrapping all queries in transactions on your behalf, and disable that setting, only using transactions explicitly when you want them.

### Client-side latency

> To reduce the roundtrip latency associated with making requests from your client to Materialize, make your requests as physically close to your Materialize region as possible. E.g. if you use the AWS us-east-1 region for Materialize, ideally your client server would also be running in AWS us-east-1. The closer geographically to the region, the lesser the network latency.

Reviewer: ```suggestion
To minimize the roundtrip latency associated with making requests from your client to Materialize, make your requests as physically close to your Materialize region as possible. For example, if you use the AWS `us-east-1` region for Materialize, your client server would ideally also be running in AWS `us-east-1`.
```
